Description: Fix truncation of large block sizes in disk space reporting
 Free and used block counts were copied into a int32_t value which
 led to truncation on systems with large storage.
Author: Mario Lang <mlang@debian.org>
Bug-Debian: http://bugs.debian.org/694623
Forwarded: no
Reviewed-By: Simon Kainz <simon.kainz@tugraz.at>
Last-Update: 2012-11-28

--- ganglia-3.3.8.orig/libmetrics/linux/metrics.c
+++ ganglia-3.3.8/libmetrics/linux/metrics.c
@@ -1174,9 +1174,6 @@ int remote_mount(const char *device, con
 float device_space(char *mount, char *device, double *total_size, double *total_free)
 {
    struct statvfs svfs;
-   uint32_t blocksize;
-   uint32_t free;
-   uint32_t size;
    /* The percent used: used/total * 100 */
    float pct=0.0;
 
@@ -1188,14 +1185,11 @@ float device_space(char *mount, char *de
       return pct;
    }
 
-   free = svfs.f_bavail;
-   size  = svfs.f_blocks;
-   blocksize = svfs.f_bsize;
    /* Keep running sum of total used, free local disk space. */
-   *total_size += size * (double) blocksize;
-   *total_free += free * (double) blocksize;
+   *total_size += svfs.f_blocks * (double) svfs.f_bsize;
+   *total_free += svfs.f_bavail * (double) svfs.f_bsize;
    /* The percentage of space used on this partition. */
-   pct = size ? ((size - free) / (float) size) * 100 : 0.0;
+   pct = svfs.f_blocks ? ((svfs.f_blocks - svfs.f_bavail) / (float) svfs.f_blocks) * 100 : 0.0;
    return pct;
 }
 
